### PR TITLE
Fix bracket preserving zero boundaries

### LINKED ISSUE #597: Fix bracket.js returning wrong initial boundary when a0 or b0 is 0
## Goal

When `bracket()` exhausts its iteration budget without finding a sign change and the caller passed `a0 = 0` (or `b0 = 0`), it should return `[a0, b0]` — i.e. `[0, b0]`. Instead it returns `[a, b0]` (the post-loop drifted value of `a`) because `a0 || a` evaluates `a` when `a0` is falsy.

This is a silent wrong-result bug: the function contracts to return the original boundaries on non-convergence, but the nullish-coalescing intent was accidentally implemented with `||` (falsy coercion) instead of `??` (nullish coalescing), breaking the zero-boundary case.

## Scope

- `src/algorithms/bracket.js` line 70: change `return [a0 || a, b0 || b]` to `return [a0 ?? a, b0 ?? b]`.
- `test/algorithms.js` "should return the specified boundaries if root was not found": extend the test to cover a case where `b0 = 0` and confirm the returned boundary is `0`, not the drifted `b`.

## Acceptance Criteria

- `bracket(t => Math.exp(-t) + 1, 0, 2)[0] === 0` continues to pass (regression).
- A new test case confirms `bracket(f, -3, 0)[1] === 0` (symmetric `b0=0` case).
- All existing algorithm tests pass.

## Out of Scope

- Changes to any other algorithm or distribution file.
- Changes to the convergence logic of `bracket()` itself.

### PR DESCRIPTION
**Pull request by yzhkali**
## Summary
- preserve zero-valued initial boundaries when `bracket()` does not find a root
- add regression coverage for both `a0 === 0` and `b0 === 0` non-convergence cases

Closes #597

## Tests
- `./node_modules/.bin/cross-env NODE_ENV=test ./node_modules/.bin/_mocha --require @babel/register test/algorithms.js`

### CODE REVIEW

**Review by synesenom — CHANGES_REQUESTED** [review]
The fix itself is correct and the test change is well-reasoned — one blocking item before merge.

## Required (P2)

**Missing `CHANGELOG.md` entry.**
Per project convention, every user-visible bug fix must have a bullet under `## [Unreleased] → ### Fixed`. The existing entry at that section covers earlier `bracket` fixes (#589) but not this one. Please add something like:

> `bracket()` now correctly returns the caller's initial `[a0, b0]` when no root is found, even when either boundary is `0`; previously `0 || <expanded>` silently returned the loop's last expanded value (#597).

---

## Informational (P3, no action required)

**Why the old test was accidentally passing.**
With `f(t) = exp(-t) + 1` and `a0=0, b0=2`, the loop always took the `|fa| > |fb|` branch (extending `b` rightward only), leaving `a` stuck at `0`. So `0 || a` evaluated to `0 || 0 = 0`, masking the bug entirely. Replacing it with `f(t) = 1` forces both-sided expansion, moves `a` away from zero, and correctly exposes the falsy-coercion. Good catch.

**Asymmetric expansion branch no longer covered in the "no root found" case.**
`f(t) = 1` exclusively triggers the equal-values `else` branch inside the loop. Since the fix is purely at the `return` statement (not inside the loop), this doesn't create a regression gap for the reported bug. Worth noting in case a future loop-level change is made.

**Pre-existing sign issue on `bracket.js:62` (out of scope for this PR).**
The `else` branch clamps `b` with `max + deltaB`, while the asymmetric branch on line 54 correctly uses `max - deltaB`. For an open upper support boundary this could allow `b` to slightly exceed `max`. Not introduced by this PR — worth a separate follow-up issue.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by synesenom — CHANGES_REQUESTED** [review]
## Required (P2 — blocking)

- [ ] [docs] `src/algorithms/bracket.js:20` — The `@return` JSDoc says "NaN otherwise", but NaN is only returned for the degenerate-input guard (`a0 === b0`, line 25). When the search loop exhausts without finding a bracket, the function returns `[a0, b0]` (an array), not NaN. Since this PR specifically addresses what is returned on the "no bracket found" path, please update the JSDoc to reflect the two distinct failure cases:

  ```js
  * @return {(number[]|number)} Array containing the bracket `[a, b]` around the root if
  *   a sign change is found, the original `[a0, b0]` if the search exhausts MAX_ITER
  *   without converging, or NaN if `a0 === b0`.
  ```

## Informational (non-blocking)

- [tests] `test/algorithms.js:44` — The new regression assertion uses a compound `assert(reverseBracket[0] === -2 && reverseBracket[1] === 0)`. If only one boundary is wrong, Mocha reports `AssertionError: false` with no indication of which element failed. Two `assert.strictEqual` calls would give more precise failure output — especially worth doing here since `b0 === 0` is the exact regression case this PR targets.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/algorithms/bracket.js
Comment (synesenom) [inline src/algorithms/bracket.js:1]: **P2 — blocking.** The `@return` JSDoc at line 20 says "NaN otherwise" but NaN is only returned for the `a0 === b0` guard (line 25). When the loop exhausts MAX_ITER without finding a bracket it returns `[a0, b0]`, not NaN. Please update line 20 to:

```js
 * @return {(number[]|number)} Array `[a, b]` containing a bracket around the root if a sign change is
 *   found, the original `[a0, b0]` if the search exhausts MAX_ITER iterations without converging, or
 *   NaN if `a0 === b0`.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/algorithms.js [hunk 1]
```diff
@@ -36,11 +36,18 @@ describe('algorithms', () => {
 
     it('should return the specified boundaries if root was not found', () => {
       const bracket = algorithms.bracket(
-        t => Math.exp(-t) + 1,
+        t => 1,
         0,
         2
       )
       assert(bracket[0] === 0 && bracket[1] === 2)
```
Comment (synesenom) [inline test/algorithms.js]: **P3 — informational.** Compound `assert(a && b)` reports `AssertionError: false` with no hint of which boundary failed. Splitting gives precise failure output:

```suggestion
      assert.strictEqual(bracket[0], 0)
      assert.strictEqual(bracket[1], 2)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

File: test/algorithms.js [hunk 2]
```diff
@@ -36,11 +36,18 @@ describe('algorithms', () => {
...
+
+      const reverseBracket = algorithms.bracket(
+        t => 1,
+        -2,
+        0
+      )
+      assert(reverseBracket[0] === -2 && reverseBracket[1] === 0)
```
Comment (synesenom) [inline test/algorithms.js]: Same — split the compound assertion so a failure identifies which boundary is wrong:

```suggestion
      assert.strictEqual(reverseBracket[0], -2)
      assert.strictEqual(reverseBracket[1], 0)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by synesenom — COMMENTED** [review]
Inline suggestions for the two findings from the REQUEST_CHANGES review.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by yzhkali** [comment]
Thanks, updated the `bracket()` return JSDoc to distinguish the MAX_ITER exhaustion path from the `a0 === b0` NaN guard, and split the boundary assertions into `assert.strictEqual` calls. The changelog entry for #604 is already present under Unreleased / Fixed.

**Review by synesenom — APPROVED** [review]
LGTM — all quality checks passed.

Fixes the classic JS falsy-zero bug (`a0 || a` evaluating to `a` when `a0 === 0`); the updated JSDoc correctly distinguishes the three return cases (`[a, b]` on success, `[a0, b0]` on exhaustion, `NaN` on degenerate input), and the split `assert.strictEqual` calls make regression failures self-describing.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by synesenom** [comment]
> Thanks, updated the `bracket()` return JSDoc to distinguish the MAX_ITER exhaustion path from the `a0 === b0` NaN guard, and split the boundary assertions into `assert.strictEqual` calls. The changelog entry for #604 is already present under Unreleased / Fixed.

Thanks for your contribution!
